flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -350,13 +340,15 @@ public CheckpointCoordinator(
                         this.minPauseBetweenCheckpoints,
                         this.pendingCheckpoints::size,
                         this.checkpointsCleaner::getNumberOfCheckpointsToClean);
+
         this.cachedTasksById =
-                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(tasksToWaitFor.length) {
+                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(
+                        attemptMappingProvider.getNumberOfTasks()) {
 
                     @Override
                     protected boolean removeEldestEntry(
                             Map.Entry<ExecutionAttemptID, ExecutionVertex> eldest) {
-                        return size() > CheckpointCoordinator.this.tasksToWaitFor.length;
+                        return size() > attemptMappingProvider.getNumberOfTasks();
```

Comment:
Why is it the number of all tasks and not just the `tasksToWaitFor`? I have the feeling that the code before was not correct when the number of tasks in the different states can change, right?

Reply:
Yes, since the incomplete tasks' metric is reported after its checkpoint is declined, if new checkpoints occur and `tasksToWaitFor` changed between the first checkpoint get declined and the metrics are reported, some tasks might be missed from the cache.  Thus it would be better to always consider all the tasks.